### ISSUE TITLE
"poetry lock" is not idempotent

### ISSUE THREAD

**Issue by sfermigier** [sdispater/poetry#782]
- [x] I am on the [latest](https://github.com/sdispater/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/sdispater/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: Mac OS X 10.13.6
- **Poetry version**: 0.12.10
- **Link of a [Gist](https://gist.github.com/) with the contents of your pyproject.toml file**: https://gist.github.com/sfermigier/c3c2cdc39836bba55427e07c776f77e2

## Issue

When running repeatedly `poetry lock` (or `poetry update`) on the same repo, the `poetry.lock` changes (with period 2):

```
-rw-r--r--    1 fermigier  staff  90098 Jan  7 16:00 poetry.lock.1
-rw-r--r--    1 fermigier  staff  89608 Jan  7 16:00 poetry.lock.2
-rw-r--r--    1 fermigier  staff  90098 Jan  7 16:01 poetry.lock.3
-rw-r--r--    1 fermigier  staff  89608 Jan  7 16:01 poetry.lock.4
-rw-r--r--    1 fermigier  staff  90098 Jan  7 16:02 poetry.lock.5
-rw-r--r--    1 fermigier  staff  89608 Jan  7 16:03 poetry.lock.6
```
(`poetry.lock.[135]` and `poetry.lock.[246]` are identical).

**Comment by dmohns** [comment]
I am not able to reproduce this behavior (using `Poetry 0.12.11` though).

On a different note: Why would expect `poetry lock` to be idempotent in the first place? In case something changes in the remote repositories we cannot expect two consecutive runs of `poetry lock` to produce the same result. 
In other words `poetry lock` is a function of the (current) timestamp and as such will never have a chance of being idempotent.

**Comment by sfermigier** [comment]
@dmohns "Why would expect poetry lock to be idempotent in the first place? In case something changes in the remote repositories we cannot expect two consecutive runs of poetry lock to produce the same result."

-> I'm talking about running `poetry lock` two times in a row, with no changes having taken place on the remote repositories.

**Comment by sfermigier** [comment]
Still not "idempotent" (in the sense I've defined above), and using Poetry 0.12.11:

https://gist.github.com/sfermigier/b753b6fb439a77b7a2d03eacc93aa587

**Comment by dmohns** [comment]
I did manage to finally produce (reproduce is really the wrong word here, see below 😂 ) this error on my machine also. I reran `poetry lock` around 20 times and one time out of those attempts it produced the faulty `poetry.lock`.

Examining the corresponding files and tree's it looks like in this scenario there is a problem with parsing extras of `requests`. Here I have (in `poetry.lock`)
```
[[package]]
category = "main"
description = "Python HTTP for Humans."
name = "requests"
optional = false
python-versions = ">=2.7, !=3.0.*, !=3.1.*, !=3.2.*, !=3.3.*"
version = "2.21.0"

[package.dependencies]
certifi = ">=2017.4.17"
chardet = ">=3.0.2,<3.1.0"
idna = ">=2.5,<2.9"
urllib3 = ">=1.21.1,<1.25"
```
while in my 19 other attemps I have
```
[[package]]
category = "main"
description = "Python HTTP for Humans."
name = "requests"
optional = false
python-versions = ">=2.7, !=3.0.*, !=3.1.*, !=3.2.*, !=3.3.*"
version = "2.21.0"

[package.dependencies]
certifi = ">=2017.4.17"
chardet = ">=3.0.2,<3.1.0"
cryptography = ">=1.3.4"
idna = [">=2.5,<2.9", ">=2.0.0", ">=2.5,<2.9", ">=2.0.0"]
pyOpenSSL = ">=0.14"
urllib3 = ">=1.21.1,<1.25"
```

Additionally, I remember I saw some `setup.cfg` parsing errors/warnings on my initial runs. My guess is that those problems are connected as the missing dependencies arise from `requests`'s `setup.cfg`. Unfortunately, I lost the logs of this and I am now back in the situation where I cannot reproduce it ...
@sfermigier Do you remember seeing similar warnings aswell?

**Comment by rafalkrupinski** [comment]
I think `poetry lock` should never change without first changing pyptoject.toml or calling `poetry update`. In mature projects you never change version of a dependency without a reasonable need, careful consideration and testing, and definitely you wouldn't want to update all dependencies at once even if it's a change of a micro version. If `poetry.lock` has all dependencies listed, it should be left alone.

**Comment by sdispater** [comment]
This should be fixed in the latest `0.12.12` release

**Comment by rafalkrupinski** [comment]
Can you explain the expected behaviour after the fix, please?

I thought there would be no change in the below scenario:
```bash
bash-3.2$ git status|grep 'pyproject\|poetry'
bash-3.2$ poetry lock
Updating dependencies
Resolving dependencies... (1.1s)

Writing lock file
bash-3.2$ git status|grep 'pyproject\|poetry'
        modified:   poetry.lock
bash-3.2$ 
```
